packages/arb-token-bridge-ui/src/components/TransferPanel/TransferPanel.tsx

```diff
@@ -675,324 +621,278 @@ export function TransferPanel() {
-                    accountType: 'EOA',
-                    network: l2NetworkName,
-                    amount: Number(amount)
-                  })
-                }
-              },
-              onTxError
-            }
+      // token approval
+      if (selectedToken) {
+        const isTokenApprovalRequired =
+          await bridgeTransferStarter.requiresTokenApproval({
+            amount: amountBigNumber,
+            signer,
+            destinationAddress
           })
-        } else {
-          await latestEth.current.withdraw({
-            amount: utils.parseUnits(amount, nativeCurrency.decimals),
-            l2Signer,
-            txLifecycle: {
-              onTxSubmit: () => {
-                openTransactionHistoryPanel()
-                setTransferring(false)
-                clearAmountInput()
-                if (!isSmartContractWallet) {
-                  trackEvent('Withdraw', {
-                    assetType: 'ETH',
-                    accountType: 'EOA',
-                    network: l2NetworkName,
-                    amount: Number(amount)
-                  })
-                }
-              },
-              onTxError
-            }
+        if (isTokenApprovalRequired) {
+          const userConfirmation = await tokenAllowanceApproval()
+          if (!userConfirmation) return false
+
+          if (isSmartContractWallet && transferType.includes('withdrawal')) {
+            showDelayInSmartContractTransaction()
+          }
+          await bridgeTransferStarter.approveToken({
+            signer
           })
         }
       }
+
+      // show a delay in case of SCW because tx is executed in an external app
+      if (isSmartContractWallet) {
+        showDelayInSmartContractTransaction()
+
+        trackEvent(isDepositMode ? 'Deposit' : 'Withdraw', {
+          tokenSymbol: selectedToken?.symbol,
+          assetType: 'ERC-20',
+          accountType: 'Smart Contract',
+          network: l2NetworkName,
+          amount: Number(amount)
+        })
+      }
+
+      // finally, call the transfer function
+      const transfer = await bridgeTransferStarter.transfer({
+        amount: amountBigNumber,
+        signer,
+        destinationAddress
+      })
+
+      // transaction submitted callback
+      onTxSubmit(transfer)
     } catch (ex) {
       console.log(ex)
     } finally {
       setTransferring(false)
     }
   }
 
+  const onTxSubmit = async (bridgeTransfer: BridgeTransfer) => {
+    const l2NetworkName = getNetworkName(parentChain.id)
+    if (!isSmartContractWallet) {
+      trackEvent(isDepositMode ? 'Deposit' : 'Withdraw', {
+        tokenSymbol: selectedToken?.symbol,
+        assetType: selectedToken ? 'ERC-20' : 'ETH',
+        accountType: 'EOA',
+        network: l2NetworkName,
+        amount: Number(amount)
+      })
+    }
+
+    const { transferType, sourceChainTransaction } = bridgeTransfer
+    const isDeposit = transferType.includes('deposit')
```

Comment:
The reason I don't like checking for the partial string is because it restricts us to specific key naming rules, which can cause bugs. We also have `cctp` type, which can be deposit or withdrawal but this doesn't cover it.

Others may have different opinions but to me it'd be nicer to have something like:

```typescript
export type BridgeTransfer = {
  type: 'deposit' | 'withdrawal'
  asset: 'native' | 'erc20'
  status: string
  sourceChainProvider: Provider
  sourceChainTransaction: { hash: string }
  destinationChainProvider: Provider
  destinationChainTransaction?: { hash: string }
}
```

instead of deriving all this from a string